### PR TITLE
Use bitwalker Java library to parse UA

### PR DESCRIPTION
**Pull request by ebuildy**
Former version was very slow to parse UA because the regexp, this PR makes it faster because the Java library.

https://github.com/HaraldWalker/user-agent-utils

<!> Output document fields will differ than former version <!>

### CODE REVIEW

**Comment by jordansissel** [comment]
> <!> Output document fields will differ than former version <!>

I am nervous about this since it could be a major compatibility breakage. Can you add some specs based on the output of this filter from the current release (asserting values like OS, device, versions, etc) and seeing that they pass with this PR?

**Comment by ebuildy** [comment]
Yes I understand this is quite sensible, I did some tests here:

https://github.com/ebuildy/logstash-filter-useragent-diff

On 500 user-agents, there are 2 JSON files output1.json and output2.json. For me, the main difference is speed, the current implementation with regexp is a big problem for performance, whereas the Java implementation is maybe x10 faster and save lot of CPU.

BTW I released a gem if you want test =>  https://rubygems.org/gems/logstash-filter-useragent2 (config name => ``useragent2``).

Thanks you,

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,5 @@
+# 2.1.0
```
Comment (suyograo) [inline CHANGELOG.md]: We should make this 3.0.0 to indicate breaking change (difference in fields from previous versions)

File: logstash-filter-useragent.gemspec [hunk 2]
```diff
@@ -1,14 +1,15 @@
 Gem::Specification.new do |s|
 
   s.name            = 'logstash-filter-useragent'
-  s.version         = '2.0.8'
+  s.version         = '2.1.0'
```
Comment (suyograo) [inline logstash-filter-useragent.gemspec]: Bump to 3.0. See my comment above

**Comment by suyograo** [comment]
@jordansissel we should bump the version of plugin to 3.0.0, and only bundle this in 5.0.0. Also, we should branch off master before merging this in.

**Comment by suyograo** [comment]
@ebuildy any performance comparison numbers you can share?

**Comment by suyograo** [comment]
Also, see #23 for more review

**Comment by ebuildy** [comment]
Sure, I have done a fresh benchmark this morning with 100.000 random user agents:

https://github.com/ebuildy/logstash-filter-useragent-diff/blob/master/benchmark.md

BTW Any suggestion to run logstash benchmarks?

**Comment by jordansissel** [comment]
> we should bump the version of plugin to 3.0.0, and only bundle this in 5.0.0

@suyograo I am worried less about performance (faster is nice!), but if the values returned by this library are _different_ than the previous useragent filter, then folks who have things like elasticsearch search and aggregations, logstash conditionals, etc, will stop working as expected.

I'd like to see that for a given data set (some useragents from a real web log), that the old useragent filter provided the exact same values as this PR for all fields (os, os version, browser, browser version, etc). If it's not exactly the same, we'll need to figure out _if_ we want to trade a major schema change for higher performance and would also need to document upgrade warnings for users who will probably have to reindex everything through this new filter to compensate.

**Comment by ebuildy** [comment]
@jordansissel you can read at https://github.com/ebuildy/logstash-filter-useragent-diff or use plugin logstash-filter-useragent2.

**Comment by jordansissel** [comment]
@ebuildy nice work gathering that! <3

Looks like there's a lot of changes to what is returned (IE vs Internet Explorer, Safari vs Mobile Safari, Other vs Unknown, etc) :|

I'm not quite sure how to proceed, but this is enough difference that we'll need to think about how to instruct users on an upgrade path.

**Comment by suyograo** [comment]
@jordansissel @ebuildy what if we add a config option which would match the old values, plus we can document the differences in values as breaking change/release notes?

**Comment by ebuildy** [comment]
Yes i think this is the best  approach, no  break changes.

So  the  2 parser implémentations will  be there ?

**Comment by suyograo** [comment]
@ebuildy spoke with @jordansissel and we are concerned about lots of user agent data that are already indexed in Elasticsearch with the old filter. This change for example, will make their aggregations return incorrect results (for new strings). 

There are 2 approaches here:
1. Create a new filter. This may be confusing to users..
2. Add a config flag which would enable the new Java implementation. We can default to the old implementation, which would protect backward compatibility, but users can choose to turn this on.

What are your thoughts?

**Comment by suyograo** [comment]
@ebuildy would you be willing to give this java lib a go? https://github.com/ua-parser/uap-java We can measure if there are any performance improvements without compromising compatibility in values

**Comment by jordansissel** [comment]
Additional notes for discussion ---

Based on observing web logs in the wild, we found that any single agent is likely to multiple requests at the same time, meaning that we can cache the useragent parsing result for a short period and avoid the cost of parsing the user agent many times.

Given this, your test data is crafted in a way that avoids the caching we implemented in this filter. I looked at the [sample.json](https://raw.githubusercontent.com/ebuildy/logstash-filter-useragent-diff/master/sample.json) and all 500 entries are unique. In practice, it is highly unlikely that every single request has a different user agent, especially given that a single browser will request multiple resources for a single page, and a user is likely to visit multiple pages.

Summarizing, while I agree with your benchmark results showing that your patch is faster, I believe it doesn't resemble real world behavior, so I don't think it informs me of much.

Let's try a few things:

1) Try a test that more closely resembles "real" agent behavior. Have each agent parsed, say, 10 times (in a row), before moving on to the next agent to parse. [1]. This should help us show the performance impact of repetition as well as showing this patch against the performance of the current filter with caching available.
2) Try the uap-java library which, in theory, has the "nice" properties of being written in Java (so we hope it's faster) and also uses the same data source as the current filter, so the schema will be the same.

[1] When loading `reddit.com` my browser makes 53 requests for content just to display this one page, so I feel 10 is a reasonable number :)

**Comment by colinsurprenant** [comment]
would also like to note that per #13 I believe that, in general, it is totally plausible to achieve a  near 100% cache hit ratio given a reasonable cache size because there are a finite number of possible useragents and large hits grouping is definitely a reality. In that respect, any parsing-only optimization will probably not provide a significative performance impact... that does not mean that improving the parsing is not a good idea but it has to be compounded with the compatibility impact and the real-life potential performance gain.

**Comment by ebuildy** [comment]
Yep I totally agree, I choice this library because it tells me if the UA is mobile, tablet or desktop, I need this information.

**Comment by suyograo** [comment]
@ebuildy not sure if you've tried using the uap-java library, but I'd like to close this since we will not be able to merge this PR.

**Comment by colinsurprenant** [comment]
I think that the non-performance related benefits of using alternative libraries are still worth investigating. We haven't found a proper way to address @ebuildy requirement to extract more/better information that the current implementation does not provide. 

The original PR intent was described as a performance improvement and in that respect we argued that given the added caching this would not provide significant improvement.

But in the context of the actual UA parsing quality (not performance), is it worth looking at how we could/should consider such improvement? As @suyograo said, a schema change is not a good idea, should we think about versioning the filter schema and let users choose schema version/format at config time? or simply create a new filer? I'd probably go with the former option to avoid user confusion on which filter to use and just default to the current schema in this multi-schema filter. 

If we agree on this maybe @ebuildy could refactor his PR to integrate his new parsing and schema into the current filter and preserve the current parsing strategy and add this new one and add a config option to choose? 

WDYT?

**Comment by ebuildy** [comment]
Hello @colinsurprenant , you can try it here => https://rubygems.org/gems/logstash-filter-useragent2 if you want to.

About current plugin caching, it's not useful for a real web site, I receive lot of visits and spam with different UA hence very high CPU (we got 1.000.000 visits per day with 15% of spam).

BTW, I choose https://github.com/HaraldWalker/user-agent-utils and it seems this project is reaching End Of Live ;/ so it was not a very good idea ....

**Pull request closed without merging**